---
title: "Heatmap of foldchanges"
author: "Hong Jiang"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    fig_caption: yes
    theme:
      bootswatch: sandstone
      base_font:
        "Arial"
      code_font:
        google: "JetBrains Mono"
      progress-bar-bg: "orange"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  warning = FALSE,
  message = FALSE
)
```
```{r library}
library(dplyr)
library(remotes)
library(ggplot2)
library(ggstatsplot)
library(tidybiology)
library(utils)
library(datasets)
library(stats)
library(graphics)
```
## Task 4
1. What is the square root of 10?
```{r task4.1}
sqrt(10)
```
2. What is the logarithm of 32 to the base 2?
```{r task4.2}
log2(32)
```
3. What is the sum of the numbers from 1 to 1000?
```{r task4.3}
sum(c(1:1000))
```
4. What is the sum of all even numbers from 2 to 1000?
```{r task4.4}
sum(seq(2,1000,2))
```
5. How many pairwise comparisons are there for 100 genes?
```{r task4.5}
choose(100, 2)
```
6. And how many ways to arrange 100 genes in triples?
```{r task4.6}
choose(100, 3) * factorial(100)
```
## Task 5 - Using R example datasets
1. Use the R internal CO2 dataset (“data(CO2)”).
```{r task5.1}
data(CO2)
```
2. Describe briefly the content of the CO2 dataset using the help function.
```{r task5.2}
help(CO2)
```
3. What is the average and median CO2 uptake of the plants from Quebec and
Mississippi?
```{r task5.3}
CO2 |>
  group_by(Type) |>
  summarise(mean = mean(uptake), median = median(uptake))
```
4. [Optional] In the “airway” example data from Bioconductor, how many genes are
expressed in each sample? How many genes are not expressed in any sample?
## Task 6 - R Functions
1. Write a function that calculates the ratio of the mean and the median of a given
vector.
```{r task6.1}
mean_by_median <- function(v) {
  ## omit na?
  mean(v) / median(v)
}
```

2. Write a function that ignores the lowest and the highest value from a given vector and
calculate the mean.
```{r task6.2}
mean_noextreme <- function(v) {
  mean(sort(v)[c(-1:-length(v))])
}
```
3. Read about piping from here:https://r4ds.had.co.nz/pipes.html#pipes (you don’t have
to learn everything, a basic understanding of the usage is enough). Write a short
(max. 300 characters, no spaces) explanation of why, how, and when not to use
pipes.  
> - why: more clean and readable
> - how: |>
> - when not: multiple inputs or branches or pipes are too long, consider modulize instead

4. Familiarize yourself with the apply-family of functions (apply, lapply, sapply etc.)
http://uc-r.github.io/apply_family
Write a short explanation (max. 300 characters, no spaces) of why they could be
useful in your work.  
  > 1. vectorize speed up the processes on elements in the alterable objects. 
  > 2. reduce replicated work
  
## Task 7 - Basic visualization with R
1. Compare the distributions of the body heights of the two species from the
'magic_guys.csv' dataset graphically
  1. using the basic 'hist' function as well as ‘ggplot’ and ‘geom_histogram’
functions from the ggplot2 package. Optimize the plots for example by trying
several different 'breaks'. Note that ggplot2-based functions give you many
more options for changing the visualization parameters, try some of them.
```{r task7.1.1}
magic_guys <- read.csv('data/magic_guys.csv')
height2plot <- magic_guys |> group_by(species) |> select(length)
ggplot(height2plot, aes(x = length, color = species)) +
  geom_histogram() + theme_classic()
```
  2. Do the same comparison as in a. but with boxplots. If you want to use the
ggplot2-package, use the functions ‘ggplot’ and ‘geom_boxplot’.
```{r task7.1.2, fig.height=1, fig.width=1.1}
ggplot(height2plot, aes(y = length, color = species)) +
  geom_boxplot() + theme_classic()
```
  3. Save the plots with the ‘png’, ‘pdf’, and ‘svg’ formats. In which situation would
you use which file format?
```{r task7.1.3}
ggsave('figures/species_boxplot.pdf')
```
> svg: small size, for web distribution
> pdf: rasterize
> png: high quality images used for web distribution

2. Load the gene expression data matrix from the ‘microarray_data.tab’ dataset
provided in the shared folder, it is a big tabular separated matrix.
  1. How big is the matrix in terms of rows and columns?
```{r task7.2.1}
microa <- read.table('data/microarray_data.tab', sep = '\t', header = T)
dim(microa)
```
  2. Count the missing values per gene and visualize this result.
```{r task7.2.2}
nacount <- colSums(is.na(microa))
hist(nacount, breaks=20)
```
  3. Find the genes for which there are more than X% (X=10%, 20%, 50%)
missing values.
```{r task7.2.3}
napercent <- colSums(is.na(microa))/colSums(!is.na(microa))
fifty_genes <- colnames(microa)[napercent>0.5]
twenty_genes <- colnames(microa)[napercent>0.2]
ten_genes <- colnames(microa)[napercent>0.1]
```
  4. Replace the missing values by the average expression value for the particular
gene. (Note: Imputing data has to be used with caution!)
```{r task7.2.4}
microa.impute <- sapply(microa, function(x) {x[is.na(x)]=mean(x, na.rm=T)})
```
3. Visualize the data in the CO2 dataset in a way that gives you a deeper understanding
of the data. What do you see?
```{r task7.3, fig.width=6}
CO2.split <- split(CO2, CO2$Type)
lapply(CO2.split, function(x) {grouped_ggscatterstats(
  data             = x,
  x                = conc,
  y                = uptake,
  type             = "parametric",
  grouping.var     = Treatment,
  ggplot.component = list(geom_rug(sides = "b"))
)})
```
follow similar linear relation
## Task 8
1. Install the Tidybiology package, which includes the data ‘chromosome’ and ‘proteins’
devtools::install_github("hirscheylab/tidybiology")
  1. Extract summary statistics (mean, median and maximum) for the following
variables from the ‘chromosome’ data: variations, protein coding genes, and
miRNAs. Utilize the tidyverse functions to make this as simply as possible.
```{r task8.1.1}
data(chromosome)
select(chromosome, c(variations, protein_codinggenes, mi_rna)) |> summary()
```
  2. How does the chromosome size distribute? Plot a graph that helps to visualize this by using ggplot2 package functions.
```{r task8.1.2}
ggplot(chromosome, aes(x = id, y=basepairs)) +
  geom_bar(stat="identity") + theme_classic()
```
  3. Does the number of protein coding genes or miRNAs correlate with the length
of the chromosome? Make two separate plots to visualize these relationships.
```{r task8.1.3}
ggscatterstats(
  data             = chromosome,
  x                = basepairs,
  y                = protein_codinggenes,
  type             = "parametric"
)
ggscatterstats(
  data             = chromosome,
  x                = basepairs,
  y                = mi_rna,
  type             = "parametric"
)
```
 >Yes
 
  4. Calculate the same summary statistics for the ‘proteins’ data variables length
and mass. Create a meaningful visualization of the relationship between
these two variables by utilizing the ggplot2 package functions. Play with the
colors, theme- and other visualization parameters to create a plot that pleases
you.
```{r task8.1.4}
data(proteins)
proteins |> select(length, mass) |> summary()
ggscatterstats(
  data             = proteins,
  x                = length,
  y                = mass,
  type             = "parametric"
)
```
