ways to arrange 100 genes in triples?
Task 5 - Using R example datasets
1. Use the R internal CO2 dataset (“data(CO2)”).
2. Describe briefly the content of the CO2 dataset using the help function.
3. What is the average and median CO2 uptake of the plants from Quebec and
Mississippi?
4. [Optional] In the “airway” example data from Bioconductor, how many genes are
expressed in each sample? How many genes are not expressed in any sample?
Task 6 - R Functions
1. Write a function that calculates the ratio of the mean and the median of a given
vector.
5
This is a helpful measure to detect data with outlying values.
Note: See Reference for R language
2. Write a function that ignores the lowest and the highest value from a given vector and
calculate the mean.
3. Read about piping from here:https://r4ds.had.co.nz/pipes.html#pipes (you don’t have
to learn everything, a basic understanding of the usage is enough). Write a short
(max. 300 characters, no spaces) explanation of why, how, and when not to use
pipes.
4. Familiarize yourself with the apply-family of functions (apply, lapply, sapply etc.)
http://uc-r.github.io/apply_family
Write a short explanation (max. 300 characters, no spaces) of why they could be
useful in your work.
Task 7 - Basic visualization with R
Comment: Files are provided in the shared Google Drive folder under Course week 1.
Examples of how to use visualization functions in R and other highly useful information that
may aid you during the course can be found here https://guardacaccia.github.io/Teaching/
The nummenmaa-package mentioned within the examples can be installed with the code
below if you wish to try some of the examples yourself.
install.packages("remotes")
library(remotes)
install_url("http://emotion.utu.fi/wp-content/uploads/2019/11
/nummenmaa_1.0.tar.gz",dependencies=TRUE)
1. Compare the distributions of the body heights of the two species from the
'magic_guys.csv' dataset graphically
a. using the basic 'hist' function as well as ‘ggplot’ and ‘geom_histogram’
functions from the ggplot2 package. Optimize the plots for example by trying
several different 'breaks'. Note that ggplot2-based functions give you many
more options for changing the visualization parameters, try some of them.
b. Do the same comparison as in a. but with boxplots. If you want to use the
ggplot2-package, use the functions ‘ggplot’ and ‘geom_boxplot’.
c. Save the plots with the ‘png’, ‘pdf’, and ‘svg’ formats. In which situation would
you use which file format?
2. Load the gene expression data matrix from the ‘microarray_data.tab’ dataset
provided in the shared folder, it is a big tabular separated matrix.
a. How big is the matrix in terms of rows and columns?
b. Count the missing values per gene and visualize this result.
c. Find the genes for which there are more than X% (X=10%, 20%, 50%)
missing values.
d. Replace the missing values by the average expression value for the particular
gene. (Note: Imputing data has to be used with caution!)
3. Visualize the data in the CO2 dataset in a way that gives you a deeper understanding
of the data. What do you see?
6
Task 8
1. Install the Tidybiology package, which includes the data ‘chromosome’ and ‘proteins’
devtools::install_github("hirscheylab/tidybiology")
a. Extract summary statistics (mean, median and maximum) for the following
variables from the ‘chromosome’ data: variations, protein coding genes, and
miRNAs. Utilize the tidyverse functions to make this as simply as possible.
b. How does the chromosome size distribute? Plot a graph that helps to
visualize this by using ggplot2 package functions.
c. Does the number of protein coding genes or miRNAs correlate with the length
of the chromosome? Make two separate plots to visualize these relationships.
d. Calculate the same summary statistics for the ‘proteins’ data variables length
and mass. Create a meaningful visualization of the relationship between
these two variables by utilizing the ggplot2 package functions. Play with the
colors, theme- and other visualization parameters to create a plot that pleases
you.